libr/anal/fcn.c

```diff
@@ -1022,41 +1025,42 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 		r_list_free (list);
 	}
 	ut64 movptr = UT64_MAX; // used by jmptbl when coded as "mov reg,[R*4+B]"
-	while (addrbytes * idx < len) {
+	ut8 buf[32]; // 32 bytes is enough to hold any instruction.
+	while (addrbytes * idx < anal->opt.bb_max_size) {
 		if (anal->limit && anal->limit->to <= addr + idx) {
 			break;
 		}
 repeat:
 		if (r_cons_is_breaked ()) {
 			break;
 		}
-		if ((len - addrbytes * idx) < 5 && len == anal->opt.bb_max_size) { // TODO: use opt.bb_max_size here
-			eprintf (" WARNING : block size exceeding max block size at 0x%08"PFMT64x"\n", addr);
-			eprintf ("[+] Try changing it with e anal.bb.maxsize\n");
+		ut32 at_delta = addrbytes * idx;
+		ut64 at = addr + at_delta;
+		ret = read_ahead (anal, at, buf, sizeof (buf));
+
+		if (ret < 0) {
+			eprintf ("Failed to read\n");
 		}
-		r_anal_op_fini (&op);
-		if (isInvalidMemory (buf + addrbytes * idx, len - addrbytes * idx)) {
+		if (isInvalidMemory (buf, sizeof (buf))) {
 			FITFCNSZ ();
 			if (anal->verbose) {
 				eprintf ("Warning: FFFF opcode at 0x%08"PFMT64x "\n", addr + idx);
 			}
 			return R_ANAL_RET_ERROR;
 		}
-		if ((oplen = r_anal_op (anal, &op, addr + idx, buf + (addrbytes * idx), len - (addrbytes * idx), R_ANAL_OP_MASK_ALL)) < 1) {
+		r_anal_op_fini (&op);
+		if ((oplen = r_anal_op (anal, &op, at, buf, sizeof (buf), R_ANAL_OP_MASK_ALL)) < 1) {
```

Comment:
Maybe this `r_anal_op` call can be repeated with a larger `buf` if it fails. What do you think?